airflow/cli/cli_parser.py

```diff
@@ -17,2165 +17,35 @@
-    action="store_true",
-)
-
-# info
-ARG_ANONYMIZE = Arg(
-    ("--anonymize",),
-    help="Minimize any personal identifiable information. Use it when sharing output with others.",
-    action="store_true",
-)
-ARG_FILE_IO = Arg(
-    ("--file-io",), help="Send output to file.io service and returns link.", action="store_true"
-)
-
-# config
-ARG_SECTION = Arg(
-    ("section",),
-    help="The section name",
-)
-ARG_OPTION = Arg(
-    ("option",),
-    help="The option name",
-)
-ARG_OPTIONAL_SECTION = Arg(
-    ("--section",),
-    help="The section name",
-)
-
-# kubernetes cleanup-pods
-ARG_NAMESPACE = Arg(
-    ("--namespace",),
-    default=conf.get("kubernetes_executor", "namespace"),
-    help="Kubernetes Namespace. Default value is `[kubernetes] namespace` in configuration.",
-)
-
-ARG_MIN_PENDING_MINUTES = Arg(
-    ("--min-pending-minutes",),
-    default=30,
-    type=positive_int(allow_zero=False),
-    help=(
-        "Pending pods created before the time interval are to be cleaned up, "
-        "measured in minutes. Default value is 30(m). The minimum value is 5(m)."
-    ),
-)
-
-# jobs check
-ARG_JOB_TYPE_FILTER = Arg(
-    ("--job-type",),
-    choices=("BackfillJob", "LocalTaskJob", "SchedulerJob", "TriggererJob"),
-    action="store",
-    help="The type of job(s) that will be checked.",
-)
-
-ARG_JOB_HOSTNAME_FILTER = Arg(
-    ("--hostname",),
-    default=None,
-    type=str,
-    help="The hostname of job(s) that will be checked.",
-)
-
-ARG_JOB_HOSTNAME_CALLABLE_FILTER = Arg(
-    ("--local",),
-    action="store_true",
-    help="If passed, this command will only show jobs from the local host "
-    "(those with a hostname matching what `hostname_callable` returns).",
-)
-
-ARG_JOB_LIMIT = Arg(
-    ("--limit",),
-    default=1,
-    type=positive_int(allow_zero=True),
-    help="The number of recent jobs that will be checked. To disable limit, set 0. ",
-)
-
-ARG_ALLOW_MULTIPLE = Arg(
-    ("--allow-multiple",),
-    action="store_true",
-    help="If passed, this command will be successful even if multiple matching alive jobs are found.",
-)
-
-# sync-perm
-ARG_INCLUDE_DAGS = Arg(
-    ("--include-dags",), help="If passed, DAG specific permissions will also be synced.", action="store_true"
-)
-
-# triggerer
-ARG_CAPACITY = Arg(
-    ("--capacity",),
-    type=positive_int(allow_zero=False),
-    help="The maximum number of triggers that a Triggerer will run at one time.",
-)
-
-# reserialize
-ARG_CLEAR_ONLY = Arg(
-    ("--clear-only",),
-    action="store_true",
-    help="If passed, serialized DAGs will be cleared but not reserialized.",
-)
-
-ALTERNATIVE_CONN_SPECS_ARGS = [
-    ARG_CONN_TYPE,
-    ARG_CONN_DESCRIPTION,
-    ARG_CONN_HOST,
-    ARG_CONN_LOGIN,
-    ARG_CONN_PASSWORD,
-    ARG_CONN_SCHEMA,
-    ARG_CONN_PORT,
-]
 
+airflow_commands = core_commands
+for executor in ExecutorLoader.import_all_executor_classes(silent=True):
```

Comment:
The implication of moving the executor cli commands to the executor modules is that if the executor module fails to load, then its cli commands won't be included in Help output. I think this is actually a fair behaviour for third party executors and even the core executors that are built into airflow. However, it is not the current behavior we have today, which is that the executor cli commands will still show up in help if they are not importable, but you'll get either warnings or stacktraces when you actually try to use them (since at that point you'll hit the importing issues). We try to catch those traces and log in this [`DefaultHelpParser._check_value`](https://github.com/aws-mwaa/upstream-to-airflow/blob/689cd36ec99fc01ab74a24b08ad78c793889feae/airflow/cli/cli_config.py#L61-L97) method, but it only looks for the celery dependency (and there could be many other ways that the module fails to load).

I'm not yet personally decided on what the right approach is here. I'd appreciate input from other folks

Reply:
One way to mitigate this, in a generalized way, would be to catch import errors for the executors as we loop through them looking for CLI commands, and then log to the user that they could not be imported, which _may_ mean some cli commands are not available for those executors and the user should first fix the issues if they wish to use those commands.

@potiuk @pierrejeambrun @eladkal any thoughts?

Reply:
This is fine. We already have precedent which is similar but even worse. When we have no CeleryExecutor **configured**, running "celery" command group yields this:

```
[jarek:~/code/airflow] [airflow-3.7] fix-infinite-nofiles-from-containerd+ 2s 130 ± airflow celery            
Usage: airflow [-h] GROUP_OR_COMMAND ...

Positional Arguments:
  GROUP_OR_COMMAND

    Groups:
      celery         Celery components
      config         View configuration

.....

Optional Arguments:
  -h, --help         show this help message and exit

airflow command error: argument GROUP_OR_COMMAND: celery subcommand works only with CeleryExecutor, CeleryKubernetesExecutor and executors derived from them, your current executor: SequentialExecutor, subclassed from: BaseExecutor, see help above.
```

I found it annoying when testing some stuff but I never got to a good solution. And when I think about it - I am more annoyed by the fact that I see the command group but I cannot actually run the command than anything else. And this is more of a maintainer problem than user problem. 

And I even thik in this case your proposal combined with moving the celery executors to providers - we could get rid of the above warning and that would solve the "annoyance" I had. This way you would only see celery command when you have "celery provider" installed (once we move the executor to celery) and not when you have it configured (and I'd love it). This is IMHO much more consistent approach

I think this is also a nice way to actually finally decouple executors - this is a nice thing that the built-in executor CLIs will add new CLI commands.

We still have the documenttion generated for all the built-in executors, so even if they are missing in the main CLI that's more than enough IMHO.

Reply:
I really appreciate the time to read the thread and send a thoughtful reply @potiuk :pray: 

Unless anyone else disagrees I'll go ahead with this approach. 
There will also need to be a followup PR which removes some of the error handling we attempt to do (basically the thing that produces that message you ran into). Core airflow should not be attempting to do this, I think we agree on that, but correct me if not.

Reply:
> Core airflow should not be attempting to do this, I think we agree on that, but correct me if not.

Yes. we agree.